From f2db693d2843838eec241232cf4d50832ee953a3 Mon Sep 17 00:00:00 2001
From: Joerg Hofrichter <joerg.hofrichter@ni.com>
Date: Wed, 26 Feb 2020 13:50:08 +0100
Subject: [PATCH] mender-helpers: Don't rely on creating a hardlink ...

... in function mender_merge_bootfs_and_image_boot_files

While copying the image boot files, creating a hardlink only works if the
paths are below the same mountpoint.

Otherwise running the task do_image_sdimg results in the following error:

WARNING: /<TMPDIR>/work/<MACHINE>/<IMAGE>/1.0-r0/temp/run.do_image_sdimg.22140:1 exit 1 from 'cp -l "$file" "$destfile"'
ERROR: Function failed: do_image_sdimg (log file is located at /<TMPDIR>/work/<MACHINE>/<IMAGE>/1.0-r0/temp

A case where this happens is when DEPLOY_DIR (or DEPLOY_DIR_IMAGE) points
to a path which is on a different mount point than the WORKDIR.

Changelog: none

Signed-off-by: Joerg Hofrichter <joerg.hofrichter@ni.com>
---
 meta-mender-core/classes/mender-helpers.bbclass | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/meta-mender-core/classes/mender-helpers.bbclass b/meta-mender-core/classes/mender-helpers.bbclass
index c9d7131..23358fb 100644
--- a/meta-mender-core/classes/mender-helpers.bbclass
+++ b/meta-mender-core/classes/mender-helpers.bbclass
@@ -286,7 +286,8 @@ mender_merge_bootfs_and_image_boot_files() {
             if [ -e "$destfile" ]; then
                 bbfatal "$destfile already exists in boot partition. Please verify that packages do not put files in the boot partition that conflict with IMAGE_BOOT_FILES."
             fi
-            cp -l "$file" "$destfile"
+            # create a hardlink if possible (prerequisite: the paths are below the same mount point), do a normal copy otherwise
+            cp -l "$file" "$destfile" || cp "$file" "$destfile"
         done
     done
 }
-- 
2.17.1

